50 create email verification

jublebi opened the pull request.

Commit:
Started with email forgot password

But mailhog seems to be not connected/reachable from quarkus container...but it is on local machine.

[skip ci]

Commit:
Implemented forgot password function

It now works to send a request to the backend which will create a new token to change the password. Normally, this token would be sent via email which is already implemented. But something in the config seems not to work. Therefore, we currently also log the token within the backend to copy it for testing purposes into the frontend.

rmzzz commented on src/main/java/at/aau/ase/cl/api/UserResource.java:
```diff
@@ -109,4 +121,62 @@ public Response login(@Valid LoginRequest loginRequest) {
+    @Path("/reset-password")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response resetPassword(ResetPasswordPayload payload) {
+        System.out.println(payload.token);
```
please replace stdout with logging

rmzzz commented on src/main/java/at/aau/ase/cl/api/UserResource.java:
```diff
@@ -15,15 +18,24 @@
 import org.eclipse.microprofile.openapi.annotations.media.Schema;
 import org.eclipse.microprofile.openapi.annotations.responses.APIResponse;
 import org.mindrot.jbcrypt.BCrypt;
+import org.slf4j.Logger;
```
we use `io.quarkus.logging.Log` in other classes also in this project

rmzzz commented on src/main/java/at/aau/ase/cl/service/ResetPasswordService.java:
```diff
@@ -0,0 +1,54 @@
+            resetPasswordEntity.userId = userId;
+            resetPasswordEntity.token = token;
+
+            entityManager.persist(resetPasswordEntity);
```
usually you don't need the entity manager for Panache entities, just `.persist()` or `.persistAndFlush()` should do it

rmzzz commented on src/main/java/at/aau/ase/cl/api/UserResource.java:
```diff
@@ -109,4 +121,62 @@ public Response login(@Valid LoginRequest loginRequest) {
+            return Response.status(Response.Status.NOT_FOUND).entity("User not found").build();
+        }
+
+        UUID resetToken = UUID.randomUUID();
```
Resource class should only dispatch calls to services. The logic should be done in Service if it is cross model logic or in Models if it fits the scope of single Model

rmzzz replied:
`UUID resetToken = UUID.randomUUID();` produces UUID v4 which is not considered secure, see e.g. https://versprite.com/blog/universally-unique-identifiers/
we could add more random bits, but then the token could not be stored as UUID anymore.

rmzzz commented on src/main/java/at/aau/ase/cl/api/UserResource.java:
```diff
@@ -109,4 +121,62 @@ public Response login(@Valid LoginRequest loginRequest) {
+        resetPasswordService.savePasswordResetToken(user.id, resetToken);
+        System.out.println(resetToken);
+
+        String resetLink = "http://localhost:8080/user/reset-password?token=" + resetToken;
```
the resetLink or at least its base url should be configurable, otherwise the service will not work in prod

rmzzz commented on src/main/java/at/aau/ase/cl/api/UserResource.java:
```diff
@@ -109,4 +121,62 @@ public Response login(@Valid LoginRequest loginRequest) {
     }
+
+    @POST
+    @Path("/forgot-password")
```
`POST /user/forgot-password`
`POST /user/reset-password`
doesn't is look like RPC style? maybe we could "rephrase" it in REST style?

rmzzz commented on src/main/java/at/aau/ase/cl/model/ResetPasswordEntity.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.UUID;
+
+@Entity
+@Table(name = "reset_password_tokens")
```
please prefix all tables with `user_` prefix to indicate they belong to user service

rmzzz commented on src/main/java/at/aau/ase/cl/service/ResetPasswordService.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Transactional
+    public void invalidateToken(ResetPasswordEntity resetPasswordEntity) {
+        ResetPasswordEntity managedEntity = entityManager.merge(resetPasswordEntity);
```
update before delete? can we just delete by unique field?

rmzzz commented on src/test/java/at/aau/ase/cl/MailhogConfigTest.java:
```diff
@@ -0,0 +1,28 @@
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+
+@QuarkusTest
+public class MailhogConfigTest {
```
is it mailer test or is it mailhog lerning test? if it is learningtest, i.e. for local execution only, please add `@Disabled("learning test")` annotation to avoid execution in automated tests

rmzzz commented on src/main/java/at/aau/ase/cl/util/JWT_Util.java:
```diff
@@ -21,7 +21,8 @@ public static String generateToken(String userId, String username, String role)
                 .groups(new HashSet<>(Arrays.asList("USER")))
```
`new HashSet<>(Arrays.asList("USER"))` -> same can be achieved with `Set.of("USER")`

rmzzz commented on src/main/java/at/aau/ase/cl/model/ResetPasswordEntity.java:
```diff
@@ -0,0 +1,32 @@
+    @Column(name = "used", nullable = false)
+    public boolean used = false;
+
+    public static ResetPasswordEntity findUnusedTokenByUserId(UUID userId) {
```
method `findUnusedTokenByUserId ` seems to be not used

rmzzz reviewed (changes requested):
a couple of minor issues and smells,
but the configurability of the reset link issue is a blocker for deploying in prod.

Commit:
Use Mailpit for Email

instead of mailhog we now use mailpit which simulates the smtp server and enables us to get emails in dev mode.

Commit:
Incorporated review

also additional changes which were stated in the meeting are now fixed. For instance the frontend user object does not contain a password anymore.

Commit:
Incorporated review

also additional changes which were stated in the meeting are now fixed. For instance the frontend user object does not contain a password anymore.

rmzzz commented on src/main/java/at/aau/ase/cl/api/UserResource.java:
```diff
@@ -62,7 +69,7 @@ public Response updatePassword(@PathParam("id") UUID id, @Valid PasswordPayload
 
         String hashedNewPassword = BCrypt.hashpw(passwordPayload.newPassword, BCrypt.gensalt());
```
password hashing concern could be moved to a single place, because now it is here and in password reset. If we have to change hashing then we will have to do change in multiple places.

rmzzz commented on src/main/java/at/aau/ase/cl/mapper/UserMapper.java:
```diff
@@ -11,5 +12,8 @@ public interface UserMapper {
+    @Mapping(target = "password", ignore = true)
+    User mapWithoutPassword(UserEntity userEntity);
+
     User map(UserEntity userEntity);
```
is there any use case when we need to map entity to an api dto with password?

rmzzz commented on src/main/java/at/aau/ase/cl/service/ResetPasswordService.java:
```diff
@@ -0,0 +1,86 @@
+
+        var user = userService.getUserById(resetTokenEntity.userId);
+        if (user == null) {
+            throw new NotFoundException("User not found");
```
is it ever possible to have missing user from ResetTokenEntity. if it is unexpected, but we'd like to avoid NPE nevertheless, I'd suggest to throw internal server error (500) instead to indicate the bug in system rather then client error (4xx).

rmzzz reviewed (approved):
thanks for fixes! 
it looks much better now!

- consider to fix the minor issues I found;
- consider to fix sonar issues as well;
- and please don't forget to add migration script for new entity, e.g. `V1.1.0__reset_password_token.sql` (note: you should not modify migration script already pushed in main, because this would break automatic migration).

then, imho, you can merge without additional review round.

Commit:
Fixed sonar issues

Commit:
Fixed sonar issues

Commit:
Added db migrations script

Commit:
Added mailpit in production

Commit:
fixed tests

broke because mailpit production configuration overwrites test
